libcontainer/dmz/dmz_linux.go

```diff
@@ -6,21 +6,32 @@ package dmz
 import (
 	"bytes"
 	"debug/elf"
-	_ "embed"
+	"embed"
 	"os"
 
 	"github.com/sirupsen/logrus"
 )
 
-// Try to build the runc-dmz binary. If it fails, replace it with an empty file
-// (this will trigger us to fall back to a clone of /proc/self/exe). Yeah, this
-// is a bit ugly but it makes sure that weird cross-compilation setups don't
-// break because of runc-dmz.
+// Try to build the runc-dmz binary on "go generate". If it fails (or
+// libcontainer is being imported as a library), the embed.FS will not contain
+// the file, which will then cause us to fall back to a clone of
+// /proc/self/exe.
 //
-//go:generate sh -c "make -B runc-dmz || echo -n >runc-dmz"
-//go:embed runc-dmz
+// There is an empty file called runc-dmz-dummy-file.txt in libcontainer/dmz in
+// order to work around the restriction that go:embed requires at least one
+// file to match the pattern.
+//
+//go:generate sh -c "make -B runc-dmz"
+//go:embed runc-dmz*
```

Comment:
Can we just embed a directory for simplicity?
https://pkg.go.dev/embed#FS.ReadDir

Reply:
We could, but we still need a dummy file. If you have an empty directory, you just get:

```
libcontainer/dmz/dmz_linux.go:24:12: pattern binary: cannot embed directory binary: contains no embeddable files
```

Reply:
Also `ReadDir` and doing logic based on it is more complicated than using `ReadFile` (if the file is not embedded we get `nil, someError` -- and we can safely ignore the error).

Reply:
This is how it looks with a directory. I guess it is a little nicer in some aspects, but it's basically the same thing:

```diff
diff --git a/libcontainer/dmz/Makefile b/libcontainer/dmz/Makefile
index 3c30db243d91..959529c5fd91 100644
--- a/libcontainer/dmz/Makefile
+++ b/libcontainer/dmz/Makefile
@@ -14,6 +14,6 @@ else
 	CFLAGS += -DRUNC_USE_STDLIB
 endif
 
-runc-dmz: _dmz.c
+binary/runc-dmz: _dmz.c
 	$(CC) $(CFLAGS) -static -o $@ $^
 	$(STRIP) -gs $@
diff --git a/libcontainer/dmz/.gitignore b/libcontainer/dmz/binary/.gitignore
similarity index 100%
rename from libcontainer/dmz/.gitignore
rename to libcontainer/dmz/binary/.gitignore
diff --git a/libcontainer/dmz/binary/dummy-file.txt b/libcontainer/dmz/binary/dummy-file.txt
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/libcontainer/dmz/binary/runc-dmz b/libcontainer/dmz/binary/runc-dmz
new file mode 100755
index 000000000000..18eda9255e75
Binary files /dev/null and b/libcontainer/dmz/binary/runc-dmz differ
diff --git a/libcontainer/dmz/dmz_linux.go b/libcontainer/dmz/dmz_linux.go
index 12f9709a269a..f2b0e8daed7d 100644
--- a/libcontainer/dmz/dmz_linux.go
+++ b/libcontainer/dmz/dmz_linux.go
@@ -6,21 +6,32 @@ package dmz
 import (
 	"bytes"
 	"debug/elf"
-	_ "embed"
+	"embed"
 	"os"
 
 	"github.com/sirupsen/logrus"
 )
 
-// Try to build the runc-dmz binary. If it fails, replace it with an empty file
-// (this will trigger us to fall back to a clone of /proc/self/exe). Yeah, this
-// is a bit ugly but it makes sure that weird cross-compilation setups don't
-// break because of runc-dmz.
+// Try to build the runc-dmz binary on "go generate". If it fails (or
+// libcontainer is being imported as a library), the embed.FS will not contain
+// the file, which will then cause us to fall back to a clone of
+// /proc/self/exe.
 //
-//go:generate sh -c "make -B runc-dmz || echo -n >runc-dmz"
-//go:embed runc-dmz
+// There is an empty file called runc-dmz-dummy-file.txt in libcontainer/dmz in
+// order to work around the restriction that go:embed requires at least one
+// file to match the pattern.
+//
+//go:generate sh -c "make -B binary/runc-dmz"
+//go:embed binary
+var runcDmzFs embed.FS
+
+// The contents of runc-dmz.
 var runcDmzBinary []byte
 
+func init() {
+	runcDmzBinary, _ = runcDmzFs.ReadFile("binary/runc-dmz")
+}
+
 // Binary returns a cloned copy (see CloneBinary) of a very minimal C program
 // that just does an execve() of its arguments. This is used in the final
 // execution step of the container execution as an intermediate process before
```